### PR TITLE
Update host position value in every period to improve precision

### PR DESCRIPTION
**Pull request by yaochunhung**
Hi,

I modify host position in every period not only larger than host period bytes. Host position will be updated more precisely. When linux side use pcm pointer callback to get host position data then the position will be updated based on pipeline period.
I also reset host position value as zero before replay ipc_stream_pcm_params to avoid unexpected value. I saw this invalid value when I run alsa conformance test and I implement PCM pointer function to get host position value but see invalid value at the beginning. After I provide this code, the invalid value will be replaced with zero.  Please help to review this PR and any feedback are welcome. Thanks.

### CODE REVIEW

File: src/audio/host.c [hunk 1]
```diff
@@ -350,6 +349,13 @@ static void host_update_position(struct comp_dev *dev, uint32_t bytes)
 	if (!hd->no_stream_position) {
 		hd->report_pos += bytes;
 
+		/* update position every period to support host query to
+		 * get the latest value.
+		 */
+		pipeline_get_timestamp(dev->pipeline, dev, &hd->posn);
+		mailbox_stream_write(dev->pipeline->posn_offset,
+				     &hd->posn, sizeof(hd->posn));
+
```
Comment (lgirdwood) [inline src/audio/host.c:358]: @yaochunhung this is a good fix, but it looks like we can tune it by setting the hd->report_pos to align with the 1ms data size. This is probably something the driver or topology should ask for (as it has impact on power) so could be fine tuned per use case.

Comment (yaochunhung) [reply src/audio/host.c:358]: @lgirdwood  Thanks for your comment. I think this change will update data based on pipeline' period. and it will be set by different use cases(via topology). It will increase some MCPS due to updating host position information(maybe also impact power). Please advice if my opinion is incorrect. Thanks.

File: src/ipc/ipc3/handler.c [hunk 2]
```diff
@@ -299,6 +300,11 @@ static int ipc_stream_pcm_params(uint32_t stream)
 	reply.rhdr.error = 0;
 	reply.comp_id = pcm_params.comp_id;
 	reply.posn_offset = pcm_dev->cd->pipeline->posn_offset;
+
+	/* reset position value before send ipc */
+	memset(&posn, 0, sizeof(posn));
+	mailbox_stream_write(reply.posn_offset, &posn, sizeof(posn));
```
Comment (lyakh) [inline src/ipc/ipc3/handler.c:306]: Could you describe in which situations you see invalid position data being delivered to Linux?

Comment (yaochunhung) [reply src/ipc/ipc3/handler.c:306]: @lyakh yes,I saw this invalid value when I run alsa conformance test and I implement PCM pointer function to get host position value at the beginning. After I provide this code, the invalid value will be replaced with zero. Not sure if this issue could be reproducible on Intels platforms because it looks like Intel uses different ways to get position at Linux side. thanks.

File: src/audio/host.c (hunk 1 above)
Comment (lgirdwood) [reply src/audio/host.c:358]: yes, I think if we can set the update period via teh driver or topology would be the best way forward. I dont think it will have too much impact on MCPS though, but it could impact power (on Intel and maybe others) as it can write to (and wake) IP blocks.

Comment (yaochunhung) [reply src/audio/host.c:358]: @lgirdwood  Does it mean that we need to define another parameter for updated_period?
I have little bit confusion about power consumption because only pipeline_get_timestamp and mailbox_stream_write without sending ipc to AP side(send ipc is still under checking by host_period_bytes). Will it impact intel platforms? Thanks.

File: src/ipc/ipc3/handler.c (hunk 2 above)
Comment (lyakh) [reply src/ipc/ipc3/handler.c:306]: got it, thanks!

File: src/audio/host.c (hunk 1 above)
Comment (lgirdwood) [reply src/audio/host.c:358]: Sorry @yaochunhung I misread the reply. Yes we need a new flag coming from the driver i.e.
```c
/* PCM params info - SOF_IPC_STREAM_PCM_PARAMS */
struct sof_ipc_pcm_params {
	struct sof_ipc_cmd_hdr hdr;
	uint32_t comp_id;
	uint32_t flags;		/**< generic PCM flags - SOF_PCM_FLAG_ */ <<<<<<<<<< Can we ad a new flag here for continuous updates
	uint32_t reserved[2];
	struct sof_ipc_stream_params params;
} __attribute__((packed, aligned(4)));
```

Comment (yaochunhung) [reply src/audio/host.c:358]: @lgirdwood  Thanks.
I check sof_ipc_pcm_params is filled in this function sof_pcm_hw_params at linux side and without platform specific callback. Not sure if what you want is to use sof_ipc_stream_params. The sof_ipc_stream_params could be add a flag naming as continuous_update(see code below maybe a better name for it) and it also can be decided in the pcm_hw_params callback function by different platforms.
```c
struct sof_ipc_stream_params {
	struct sof_ipc_hdr hdr;
	struct sof_ipc_host_buffer buffer;
	uint32_t direction;	/**< enum sof_ipc_stream_direction */
	uint32_t frame_fmt;	/**< enum sof_ipc_frame */
	uint32_t buffer_fmt;	/**< enum sof_ipc_buffer_format */
	uint32_t rate;
	uint16_t stream_tag;
	uint16_t channels;
	uint16_t sample_valid_bytes;
	uint16_t sample_container_bytes;

	uint32_t host_period_bytes;
	uint16_t no_stream_position; /**< 1 means don't send stream position */
        uint16_t continuous_update_position; /**< 1 means continue updating stream position */  <<<<< add a flag for it
	uint16_t reserved[2];
	uint16_t chmap[SOF_IPC_MAX_CHANNELS];	/**< channel map - SOF_CHMAP_ */
} __packed;
```

Comment (yaochunhung) [reply src/audio/host.c:358]: @lgirdwood  Thanks for suggestion. I add a PR to add this flag in kernel first. https://github.com/thesofproject/linux/pull/3353/ Please help to check if it is ok. Thanks!

Comment (yaochunhung) [reply src/audio/host.c]: @lgirdwood  Do I need to modify abi.h because it looks like there is a checkpatch warning as below? If yes, how to modify it? Thanks. 
https://sof-ci.01.org/sofpr/PR5137/build11487/checkpatch/
```
WARNING: Please update ABI in accordance with http://semver.org
#72: FILE: src/include/ipc/stream.h:97:
 	uint16_t no_stream_position; /**< 1 means don't send stream position */
+	uint16_t cont_update_posn; /**< 1 means continue update stream position */
```

File: src/audio/host.c [hunk 3]
```diff
@@ -308,7 +309,6 @@ static void host_update_position(struct comp_dev *dev, uint32_t bytes)
 	struct comp_buffer *sink;
 	int ret;
 
-
```
Comment (dbaluta) [inline src/audio/host.c:311]: Please dont remove this new line here. Is not related to this patch. If you really want to remote it create another patch and explain why.

File: src/audio/host.c [hunk 4]
```diff
@@ -76,6 +76,7 @@ struct host_data {
 	uint32_t host_period_bytes;
```
Comment (dbaluta) [inline src/audio/host.c:76]: Please capture in your commit message the explanation given here:

image

Rember that a commit message should explain WHY the commit is needed. The HOW should and WHAT should be obvious from the coded

File: src/audio/host.c (hunk 3 above)
Comment (yaochunhung) [reply src/audio/host.c:311]: @dbaluta  OK. Got it. I will fix it. I remove it because there are two extra lines here.

File: src/audio/host.c (hunk 4 above)
Comment (yaochunhung) [reply src/audio/host.c:76]: @dbaluta Thanks for suggestion. I have added more information in the commit message.

**Comment by dbaluta** [comment]
@yaochunhung looks good to me. I think we also need to update the ABI minor version. @lgirdwood ?

File: src/include/ipc/stream.h [hunk 5]
```diff
@@ -94,8 +94,9 @@ struct sof_ipc_stream_params {
 
 	uint32_t host_period_bytes;
 	uint16_t no_stream_position; /**< 1 means don't send stream position */
+	uint8_t cont_update_posn; /**< 1 means continue update stream position */
```
[the remark this answers is not in the document]
Comment (yaochunhung) [reply src/include/ipc/stream.h:97]: @ranj063  I will fix it. Thanks

File: src/audio/host.c (hunk 1 above)
Comment (lgirdwood) [reply src/audio/host.c]: yes, we need to label the PR with the ABI flag - that way everyone knows there is ABI change. I've marked both PRs with ABI label now.

**Comment by lgirdwood** [comment]
@yaochunhung ABI is 3.21, this will need updated in header file.

File: src/audio/host.c (hunk 1 above)
Comment (yaochunhung) [reply src/audio/host.c]: @lgirdwood  I also update abi version to ABI-3.21. thanks

**Comment by yaochunhung** [comment]
@lgirdwood  I have updated version 3.21. please help to review it. Thanks.

File: src/audio/host.c [hunk 6]
```diff
@@ -350,6 +351,14 @@ static void host_update_position(struct comp_dev *dev, uint32_t bytes)
...
 		hd->report_pos += bytes;
 
+		/* update position every period to support host query to
+		 * get the latest value if cont_update_posn == 1.
+		 */
+		if (hd->cont_update_posn == 1) {
+			pipeline_get_timestamp(dev->pipeline, dev, &hd->posn);
+			mailbox_stream_write(dev->pipeline->posn_offset,
+					     &hd->posn, sizeof(hd->posn));
+		}
```
Comment (ujfalusi) [inline src/audio/host.c:361]: The `hd->report_pos` is not used if `hd->cont_update_posn == 1`

```c
if (!hd->no_stream_position) {
	hd->report_pos += bytes;

	if (hd->cont_update_posn == 1) {
		/*
		 * update position every period to support host query to get the
		 * latest value if cont_update_posn == 1.
		 */
		pipeline_get_timestamp(dev->pipeline, dev, &hd->posn);
		mailbox_stream_write(dev->pipeline->posn_offset,
					&hd->posn, sizeof(hd->posn));
	} else if (hd->host_period_bytes != 0 &&
		   hd->report_pos >= hd->host_period_bytes) {
		/*
		 * host_period_bytes is set to zero to disable position update
		 * by IPC for FW version before 3.11, so send IPC message to
		 * driver according to this condition and report_pos.
		 */
		hd->report_pos = 0;
		/* send timestamped position to host
		* (updates position first, by calling ops.position())
		*/
		pipeline_get_timestamp(dev->pipeline, dev, &hd->posn);
		mailbox_stream_write(dev->pipeline->posn_offset,
				&hd->posn, sizeof(hd->posn));
		ipc_msg_send(hd->msg, &hd->posn, false);
	}
}
```
But I would be tempted to move the `hd->report_pos += bytes;` to the else case as with continuous update you do not even need to increment it.

What would be the side effect if this is done unconditionally, without a flag from the host?

**Review by ujfalusi — CHANGES_REQUESTED** [review]

File: src/audio/host.c (hunk 6 above)
Comment (yaochunhung) [reply src/audio/host.c:361]: @ujfalusi  hd->report_pos is still needed when hd->cont_update_posn == 1 because when hd->cont_update_posn is set as '1', only update host position to mailbox buffer without sending ipc to notify the host.
About the side effect without a flag from the host,  
Please reference the discussion as the link https://github.com/thesofproject/sof/pull/5137#discussion_r776066511. @lgirdwood has the suggestion about it. Thanks

Comment (ujfalusi) [reply src/audio/host.c:361]: @yaochunhung, I see. so this will update the position data in between IPCs, `no_stream_position == 0` -> the position IPC is enabled.
How is the position updated in case of `no_stream_position == 1'?

**Review by kv2019i — CHANGES_REQUESTED** [review]
Thanks @yaochunhung , this looks like a good addition to the IPC interface. I have some comments on how this is implemented, please see inline.

File: src/audio/host.c (hunk 6 above)
Comment (kv2019i) [reply src/audio/host.c:361]: Please see my review comment, I think we should clean up the semantics, the old fields are quite hard to follow (and worse if we add more fields on top). I think host should still have option to prevent both mailbox and IPC updates (traffic to mailbox is non-zero cost on many platforms so if host can get position info via other means, it probably doesn't need the mailbox traffic -> the case on Intel platforms)

File: src/audio/host.c [hunk 7]
```diff
@@ -76,6 +76,7 @@ struct host_data {
 	uint32_t host_period_bytes;
 	uint16_t stream_tag;
 	uint16_t no_stream_position; /**< 1 means don't send stream position */
+	uint8_t cont_update_posn; /**< 1 means continuous update stream position */
```
[the remark this answers is not in the document]
Comment (yaochunhung) [reply src/audio/host.c:79]: @kv2019i  Thanks for suggestion but for my case I still need to send IPC but update mailbox not depend on host_period_bytes. My original problem is to update mailbox not in time (e.g. 10ms host period size) and pipeline period is 1ms. It causes that I can'
t pass alsa_conformance_test test cases(rate is out of spec).  Therefore, I want to update mailbox more frequently to get more precisive host position.   Or  can we define position_reporting as 2 for mailbox(based on pipeline period size) + IPC? Not sure which cases Intel's platform will use? Thanks.

[the remark this answers is not in the document]
Comment (ujfalusi) [reply src/audio/host.c:79]: @yaochunhung, which platform is this? If we can not provide precise position we should be setting the SNDRV_PCM_INFO_BATCH, like we do for bdw/byt and pci-tng?

[the remark this answers is not in the document]
Comment (yaochunhung) [reply src/audio/host.c:79]: @ujfalusi  mt8195, do you mean SNDRV_PCM_INFO_BATCH can be used to get more precisive data? If yes, I may check it. Thanks.

[the remark this answers is not in the document]
Comment (ujfalusi) [reply src/audio/host.c:79]: No, the flag is to indicate ALSA that the hardware is not capable of precise DMA position reporting.
If this flag is not set then it is expected that we can have access to the DMA position in fine granularity, as a reference it is only omitted for DMAengine platforms if the DMA can report with `DMA_RESIDUE_GRANULARITY_BURST', SEGMENT and DESCRIPTOR granularity will adds the BATCH flag.

If the BATCH flag is set then ALSA will interpolate the position based on elapsed time and does the correction at period elapsed times.

[the remark this answers is not in the document]
Comment (kv2019i) [reply src/audio/host.c:79]: Thanks @yaochunhung , I understand now. So you want IPC updates at a lower pace, but mailbox updates at "DSP rate" (e.g. typically every 1ms).

I think SNDRV_PCM_INFO_BATCH is probably worth a check. If you just want to keep the comformance test happy, having the host side extrapolate the position might be overall better, than having the FW write to mailbox every 1ms. 

If that is not good and we really need to expose, this requires some further thought how to best expose in the interface. 
Definitely just adding more differnet modes is one option, e.g. have one "reporting mode for DSP-rate-mailbox plus host-rate IPC", but not sure about this yet.

[the remark this answers is not in the document]
Comment (yaochunhung) [reply src/audio/host.c:79]: @ujfalusi @kv2019i  Thanks for suggestion. I try to add  SNDRV_PCM_INFO_BATCH to hw_info for alsa_conformance_test but it seems not work(rate and underrun number are the same as which without SNDRV_PCM_INFO_BATCH). I am not sure if it won't work under alsa_conformance_test only or other usages. Thanks.

**Comment by lgirdwood** [comment]
@yaochunhung any update, are we aligned kernel ?

**Comment by yaochunhung** [comment]
@lgirdwood There are still some discussion in the kernel PR https://github.com/thesofproject/linux/pull/3353 . I will check and give feedback on it. thanks.

File: src/audio/host.c (hunk 6 above)
Comment (plbossart) [inline src/audio/host.c:361]: I have tried multiple times to understand this PR, and I really don't get the idea.

If the idea is to send an IPC back to the host to update the timestamps every time a host DMA block is handled by the firmware, this is a VERY BAD IDEA.

if e.g. the host component retrieves blocks of 1ms, then you will have 1000 interrupts per second to update a timestamp, but the .pointer on the kernel side only needs this information on a period_elapsed event, or on a snd_pcm_status request.

It would be a much better idea to cache the timestamp in a memory window shared with the host and have the driver read from it.

**Review by plbossart — CHANGES_REQUESTED** [review]

File: src/audio/host.c (hunk 6 above)
Comment (yaochunhung) [reply src/audio/host.c:361]: @plbossart I don't change IPC(interrupt) in this PR. I still keep the conditional check in the else if condition below, which is still depend on host_period_bytes(period size from HOST side,not  host component period in DSP firmware side), I only want to update host position data in host component period(1ms) but doensn't send IPC to kernel. Thanks.
```c
if (!hd->no_stream_position) {
	hd->report_pos += bytes;

+	if (hd->cont_update_posn) {  
+		/*
+		 ***** update position every period to support host query to get the
+		 * latest value if cont_update_posn == 1.****
+		 */
+		pipeline_get_timestamp(dev->pipeline, dev, &hd->posn);
+		mailbox_stream_write(dev->pipeline->posn_offset,
+				     &hd->posn, sizeof(hd->posn));
+	} 
       
       if (hd->host_period_bytes != 0 &&
		   hd->report_pos >= hd->host_period_bytes) {     
		/*
		 * host_period_bytes is set to zero to disable position update
		 * by IPC for FW version before 3.11, so send IPC message to
		 * driver according to this condition and report_pos.
		 */
		hd->report_pos = 0;
		/* send timestamped position to host
		* (updates position first, by calling ops.position())
		*/
		pipeline_get_timestamp(dev->pipeline, dev, &hd->posn);
		mailbox_stream_write(dev->pipeline->posn_offset,
				     &hd->posn, sizeof(hd->posn));
		ipc_msg_send(hd->msg, &hd->posn, false);   <==Send IPC here, it depend on host_period_bytes
	}
}
```

Comment (plbossart) [reply src/audio/host.c:361]: ok, if you update the position in a mailbox that's fine, but then why do you initiate an IPC in your .pointer function in the kernel? You are not using the mailbox information, are you? I really don't see how the two sides work together.


````c
snd_pcm_uframes_t mtk_adsp_pcm_pointer(struct snd_sof_dev *sdev,
				       struct snd_pcm_substream *substream)
{
	struct snd_soc_pcm_runtime *rtd = asoc_substream_to_rtd(substream);
	struct snd_soc_component *scomp = sdev->component;
	struct snd_sof_pcm *spcm;
	struct sof_ipc_stream_posn posn;
	struct snd_sof_pcm_stream *stream;
	snd_pcm_uframes_t pos;
	int ret;

	spcm = snd_sof_find_spcm_dai(scomp, rtd);
	if (!spcm) {
		dev_warn_ratelimited(sdev->dev, "warn: can't find PCM with DAI ID %d\n",
				     rtd->dai_link->id);
		return 0;
	}

	stream = &spcm->stream[substream->stream];
	ret = snd_sof_ipc_msg_data(sdev, stream->substream, &posn, sizeof(posn));
	if (ret < 0) {
		dev_warn(sdev->dev, "failed to read stream position: %d\n", ret);
		return 0;
	}

	memcpy(&stream->posn, &posn, sizeof(posn));
	pos = spcm->stream[substream->stream].posn.host_posn;
	pos = bytes_to_frames(substream->runtime, pos);

	return pos;
}

````

Comment (yaochunhung) [reply src/audio/host.c:361]: @plbossart yes, I write host position into mailbox information on DSP firmware, so I get the information in .pointer callback function mtk_adsp_pcm_pointer. Thanks.
```c
host.c
		mailbox_stream_write(dev->pipeline->posn_offset,
				     &hd->posn, sizeof(hd->posn));
```

File: src/audio/host.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (kv2019i) [reply src/audio/host.c:79]: Sorry for late ack @yaochunhung . Ok, then having an option to have higher freq mailbox updates is probably good to have in the API. I still think a solution that would not cause this much memory window traffic would be better, but I don't think SOF interface should prevent this either.

File: src/audio/host.c (hunk 6 above)
Comment (kv2019i) [reply src/audio/host.c:361]: Ack @plbossart , the "ipc_msg_data" will read from the mailbox (not send an IPC) on all platforms. Most use the common implementation which is a call to snd_sof_dsp_mailbox_read().

So basicly this adds a new variable to the interface to configuration position updates:
  1) should position in mailbox be updated at a) DSP tick rate, b) host period rate or c) not at all
  2) should IPC be sent at a) DSP tick rate, b) host period rate, c) not at all

Sending IPC without mailbox update makes no sense, so 2 depends on 1.

On Intel platforms, we use 1c/2c  for production, but allow 1b+2b for debug.

@yaochunhung is seeing issues for production and would like the ability to configure 1a+2b (or 1a+2c ?) for mtk platforms.

Updating mailbox more often has a performance impact, but I think the SOF interface should allow this.

Comment (yaochunhung) [reply src/audio/host.c:361]: @kv2019i  Thanks for detail description. For mediatek, we want to use 1a + 2b. Thanks.

Comment (plbossart) [reply src/audio/host.c:361]: I am still confused, 'host_period_rate' means what? 
a) the ALSA period in the DDR ring buffer 
b) the size of the host buffer component in the firmware?

if this is b), then I maintain that 2b) is crazy.
if this is a), then I don't see how the precision is increased.

Comment (yaochunhung) [reply src/audio/host.c:361]: @plbossart it is a), and we update mailbox data in b) when add the codes below,
```c
host.c in DSP firmware
+	if (hd->cont_update_posn) {  
+		/*
+		 ***** update position every period to support host query to get the
+		 * latest value if cont_update_posn == 1.****
+		 */
+		pipeline_get_timestamp(dev->pipeline, dev, &hd->posn);
+		mailbox_stream_write(dev->pipeline->posn_offset,
+				     &hd->posn, sizeof(hd->posn));
+	} 
```
then it will be gotten in .pointer callback when call snd_sof_ipc_msg_data function as @kv2019i  mentioned. thanks.

Comment (kv2019i) [reply src/audio/host.c:361]: @plbossart So it's the ALSA period_size. Whenever application calls snd_pcm_avail() (or related calls), this will do a hwsync, , and end up calling pointer() SOF DSP ops. For MTK, this would not use a MMIO or info from most recently sent position IPC, but rather read the position from mailbox directly (i.e. mapped DSP memory). Main drawback I see is the high amount of SRAM write traffic on DSP side, but other than that, this should get better accuracy.

**Comment by yaochunhung** [comment]
@plbossart @kv2019i @ujfalusi  Thanks for review and comments.  Please let me know if this PR is ok or I need any improvement on it. Thanks.

File: src/audio/host.c (hunk 6 above)
Comment (plbossart) [reply src/audio/host.c:361]: ok that makes sense, but now I am confused on what we do today before this PR? Is this saying that when a period_elapsed event is signaled, the firmware does not have the latest information? 

In other words, I wonder if this 'new' behavior should have been the default behavior from the start, and if we need the new flag in the IPC structure or just an unconditional change in the firmware?

Comment (yaochunhung) [reply src/audio/host.c:361]: @plbossart As @kv2019i mentioned, there are two ways to update position information. one is (A). period_elapsed event and another is (B).application call snd_pcm_avail(or related calls). For case (A), there is no change with/without this PR. For case (B), firmware will have the latest information with this PR.  alsa_conformance_test will use both case (A) and (B).  I set a new flag "cont_update_posn" in IPC structure "sof_ipc_stream_params " so that user can decide to use this new behavior on their platforms or not. Thanks.

File: src/audio/host.c (hunk 7 above)
Comment (plbossart) [inline src/audio/host.c:79]: suggested commit message edit:

The position is currently updated by the host component when a period was consumed. The granularity of this position is too large, applications relying on timer-based scheduling may query the position at any time and conformance test suites fail.

This patch suggests a more precise position reporting, by storing the latest position in a memory window when the host component completes a transfer. This will increase the traffic to memory windows, but allow for a much smaller granularity in position updates.

In hindsight, this solution should have been the default behavior but to avoid backwards compatibility issues is added as a new capability controlled by an ABI  check. Only with a recent-enough Linux kernel will this behavior be enabled.

**Review by plbossart — CHANGES_REQUESTED** [review]
please rebase and revisit the commit message for clarify, thank you @yaochunhung

File: src/audio/host.c (hunk 7 above)
Comment (yaochunhung) [reply src/audio/host.c:79]: @plbossart  Thanks a lot for providing this commit message. I think it would be better to understand.

**Review by RanderWang — APPROVED** [review]
it is developed as ipc4 now

File: src/audio/host.c (hunk 6 above)
Comment (plbossart) [inline src/audio/host.c:361]: I am a bit confused now on the relationship between 'no_stream_position' and 'cont_update_posn'.

Is there really a hierarchy between the two, or could 'cont_update_posn' be enabled when we don't send any IPC?

Comment (yaochunhung) [reply src/audio/host.c:361]: @plbossart  good question. from the description of no_stream_position is "1 means don't send stream position".  In original design, IPC should be sent when no_stream_position is zero and satisfy the condition below, 
```c
 "if (hd->host_period_bytes != 0 &&
		    hd->report_pos >= hd->host_period_bytes)"
```
if no_stream_position is set as 1, that means host won't get stream position from DSP firmware. Maybe cont_update_posn will be no used? Thanks.

Comment (yaochunhung) [reply src/audio/host.c:361]: I wonder if there is any case that needs to update stream position from DSP firmware without IPC requirement because DSP send IPC : SOF_IPC_STREAM_POSITION to notify kernel then call ipc_period_elapsed to update stream position and call snd_sof_pcm_period_elapsed for period wakeup mode.

Comment (lgirdwood) [reply src/audio/host.c:361]: The position can be available to host in 4 ways today.
1) Host sends IPC (latency adds some variability here - not very accurate)
2) FW sends notifications to host (more accurate than 1 since its a one way message, but still subject to host latencies when receiving the notification).
3) FW writes to shared SRAM for every copy(). Accurate to 1ms (LL scheduling tick)
4) Host reads from HW position registers (Intel HW can do this today).

Comment (yaochunhung) [reply src/audio/host.c:361]: @lgirdwood  Thanks for detail explanation. I have some questions below,
Q1. Does any platform use method 1 since it is not accurate?
Q2. Does Method 3 exist in current design or it is the same as this PR do?
Q3. Can Method 3 be used along? I think we need one way to notify Host to get position?
Thanks a lot.

Comment (yaochunhung) [reply src/audio/host.c:361]: @plbossart Could you please tell me if you still have any concern or suggestion on this PR? Thanks.

Comment (plbossart) [reply src/audio/host.c:361]: I still have the concern on what happens in the no_stream_position case. @ujfalusi had a similar question.

In addition, it's not a good idea to have multiple requests to get timestamps. I would use something like this to make sure the timestamp queries are unified.

````diff
diff --git a/src/audio/host.c b/src/audio/host.c
index 8f0495723..f9014b847 100644
--- a/src/audio/host.c
+++ b/src/audio/host.c
@@ -306,6 +306,8 @@ static void host_update_position(struct comp_dev *dev, uint32_t bytes)
        struct host_data *hd = comp_get_drvdata(dev);
        struct comp_buffer *source;
        struct comp_buffer *sink;
+       bool update_mailbox = false;
+       bool send_ipc = false;
        int ret;
 
 
@@ -346,10 +348,12 @@ static void host_update_position(struct comp_dev *dev, uint32_t bytes)
                hd->local_pos = 0;
 #endif
 
+       hd->report_pos += bytes;
+       if (hd->cont_update_posn)
+               update_mailbox = true;
+
        /* Don't send stream position if no_stream_position == 1 */
        if (!hd->no_stream_position) {
-               hd->report_pos += bytes;
-
                /* host_period_bytes is set to zero to disable position update
                 * by IPC for FW version before 3.11, so send IPC message to
                 * driver according to this condition and report_pos.
@@ -361,12 +365,19 @@ static void host_update_position(struct comp_dev *dev, uint32_t bytes)
                        /* send timestamped position to host
                         * (updates position first, by calling ops.position())
                         */
-                       pipeline_get_timestamp(dev->pipeline, dev, &hd->posn);
-                       mailbox_stream_write(dev->pipeline->posn_offset,
-                                            &hd->posn, sizeof(hd->posn));
-                       ipc_msg_send(hd->msg, &hd->posn, false);
+                       update_mailbox = true;
+                       send_ipc = true;
                }
        }
+
+       if (update_mailbox) {
+               pipeline_get_timestamp(dev->pipeline, dev, &hd->posn);
+               mailbox_stream_write(dev->pipeline->posn_offset,
+                                    &hd->posn, sizeof(hd->posn));
+               if (send_ipc)
+                       ipc_msg_send(hd->msg, &hd->posn, false);
+       }
+
 }
 
 /* The host memory is not guaranteed to be continuous and also not guaranteed
````

Comment (yaochunhung) [reply src/audio/host.c:361]: @plbossart  Thanks for providing suggestion to ensure timestamp query are unified. But I am not sure if it is fine to move  the code below to the no_stream_position case(outside of the if(!hd->no_stream_position)) 
```` diff
+       hd->report_pos += bytes;
+       if (hd->cont_update_posn)
+               update_mailbox = true;
````
 because hd->report_pos  is circular and will be reset as zero under the condition. thanks.
```c
		if (hd->host_period_bytes != 0 &&
		    hd->report_pos >= hd->host_period_bytes) {
			hd->report_pos = 0;
```

Comment (plbossart) [reply src/audio/host.c:361]: I can't figure out this code, it's just convoluted for no good reason. @lgirdwood can you take a look?

Comment (yaochunhung) [reply src/audio/host.c:361]: @lgirdwood Sorry to bother. Could you please help to check it and give suggestion if any? Thanks.

Comment (lgirdwood) [reply src/audio/host.c:361]: @yaochunhung apologies, I did not see the question.
It looks like to me that we only need the 
```diff
+       if (update_mailbox) {
+               pipeline_get_timestamp(dev->pipeline, dev, &hd->posn);
+               mailbox_stream_write(dev->pipeline->posn_offset,
+                                    &hd->posn, sizeof(hd->posn));
```
We can write to the mailbox on every pipeline time tick. However, this should be a topology option "update mailbox position" flag (just like the no position update flag). Would this work for you to read from mailbox ? This would mean no IPC traffic and host could read when it pleased ?

Comment (yaochunhung) [reply src/audio/host.c:361]: @lgirdwood  Thanks for comment. I provide this patch and it looks like fine for me. I will provide a update for review. Thanks.

**Review by lgirdwood — APPROVED** [review]
@plbossart good for kernel side.
@yaochunhung this will make v2.1 since it's default state is OFF and only enabled by topology.

File: src/include/kernel/abi.h [hunk 8]
```diff
@@ -29,8 +29,8 @@
 
 /** \brief SOF ABI version major, minor and patch numbers */
 #define SOF_ABI_MAJOR 3
-#define SOF_ABI_MINOR 20
-#define SOF_ABI_PATCH 1
+#define SOF_ABI_MINOR 21
+#define SOF_ABI_PATCH 0
```
Comment (dbaluta) [inline src/include/kernel/abi.h:34]: @yaochunhung @lgirdwood  what is SOF_ABI_PATCH and why is decremented?

**Review by plbossart — APPROVED** [review]
Sorry I missed the update, thanks @lgirdwood for the ping.
